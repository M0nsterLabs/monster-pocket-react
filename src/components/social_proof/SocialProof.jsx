import React from 'react';
import ReactDOM from 'react-dom';
const Shortid = require('shortid');
import './SocialProof.less';
export default class SocialProof extends React.Component {
  static propTypes = {
    notice: React.PropTypes.oneOfType([
      React.PropTypes.array,
      React.PropTypes.object
    ]),
    width: React.PropTypes.number
  };

  static defaultProps = {
    notice             : {},
    width              : 340,
    afterRemoveDomNode : () => {
    }
  };
  state = {
    notice      : [],
    fixed       : false,
    topPosition : 0
  };
  isUpdate = false;
  refsArray = [];

  constructor (props) {
    super(props);
    if (Array.isArray(props.notice)) {
      this.state.notice = props.notice.map((item) => {
        item.type = item.type || 'default';
        item.icon = item.icon || false;
        item.hide = false;
        return item;
      });
    } else {
      let item = props.notice;
      item.type = item.type || 'default';
      item.icon = item.icon || false;
      item.hide = false;
      this.state.notice = [item];
    }
  };

  onHideNotice = (id) => {
    const noticeHTML = this.refsArray.find((element) => {
      let result = null;
      if (element.getAttribute('data-id') === id) {
        result = element;
      }
      return result;
    });
    let state = this.state;
    setTimeout(() => {
      noticeHTML.style.marginTop = (noticeHTML.offsetHeight * -1) + 'px';
    }, 1);
    state.notice = this.state.notice.map((element) => {
      let item = element;
      if (item.id === id) {
        item.hide = true;
      }
      return item;
    });
    state = this.clearNoticeObject(state);
    this.setState(state);
    noticeHTML.addEventListener('animationend', () => {
      state.notice = this.state.notice.filter((element) => {
        return element.id !== id;
      });
      this.setState(state);
      if (!this.state.notice.length) {
        this.props.afterRemoveDomNode();
        ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(this).parentNode);
      }
    });
  };

  clearNoticeObject = (state = this.state) => {
    state.notice.map((item) => {
      let element = item;
      element.className = '';
      return element;
    });
    return state;
  };

  calcMarginTop = () => {
    let totalHeight = 0;
    if (this.isUpdate) {
      this.refsArray.forEach((element) => {
        if (element.classList.contains('new-element')) {
          totalHeight += (element.offsetHeight || 0) + 20;
          element.classList.remove('new-element');
        }
      });
    }
    this.refsArray.forEach((element, i) => {
      if (totalHeight) {
        if (i === 0) {
          element.style.marginTop = (totalHeight * -1) + 'px';
        } else {
          element.style.marginTop = '20px';
        }
      } else {
        element.style.marginTop = (element.offsetHeight * -1) + 'px';
      }
      element.addEventListener('animationend', () => {
        element.style.marginTop = '20px';
        element.classList.remove('notice_show');
      });
    });
  };

  componentWillReceiveProps = (nextProps) => {
    let newProps = nextProps;
    let state = this.clearNoticeObject();
    if (Array.isArray(newProps.notice)) {
      newProps.notice.map((item) => {
        let element = item;
        element.className = 'new-element notice_show';
        element.id = Shortid.generate();
        state.notice.unshift(element);
      });
    } else {
      let element = newProps.notice;
      element.className = 'new-element notice_show';
      element.id = Shortid.generate();
      state.notice.unshift(element);
    }
    this.isUpdate = true;
    state.notice = state.notice.map((item) => {
      item.type = item.type || 'default';
      item.icon = item.icon || false;
      return item;
    });
    this.setState(state);
  };

  componentDidUpdate = () => {
    if (this.isUpdate) {
      this.calcMarginTop();
      this.isUpdate = false;
    }
  };

  componentDidMount = () => {
    this.calcMarginTop();
    this.setState({
      ...this.state,
      topPosition: ReactDOM.findDOMNode(this).offsetTop
    });
    document.addEventListener('scroll', e => {
      const scrolled = window.pageYOffset || document.documentElement.scrollTop;
      if (scrolled > this.state.topPosition && !this.state.fixed) {
        this.setState({
          ...this.state,
          fixed: true
        });
      }
      if (scrolled < this.state.topPosition && this.state.fixed) {
        this.setState({
          ...this.state,
          fixed: false
        });
      }
    });
  };

  componentWillMount = () => {
    let state = this.state;
    let notice = [];
    [...state.notice].forEach((element, i) => {
      element.id = Shortid.generate();
      if (element.className) {
        element.className += ' notice_show';
      } else {
        element.className = ' notice_show';
      }
      notice[i] = element;
    });
    state.notice = notice;
    this.setState(state);
  };

  render () {
    this.refsArray = [];
    return (
      <div className={`notice-wrapper ${this.state.fixed ? 'notice-wrapper_fixed' : ''}`} style={{width: this.props.width}}>
        {this.state.notice.map((item) => {
          const iconClass = (item.icon) ? `notice_has-icon notice_icon-${item.icon}` : '';
          return (
            <div
              className={`notice notice_type-${item.type} ${iconClass} ${(item.hide) ? 'notice_hide' : ''} ${item.className.length ? item.className : ''}`}
              key={item.id}
              style={{marginTop: '-20px'}}
              data-id={item.id}
              ref={(e) => {
                if (e) this.refsArray.push(e);
              }}
            >
              <span className="notice-text t3">{item.text}</span>
              <div className='notice__closeBlock'>
                <div
                  className='notice__closeBlock__closeArea'
                  onClick={this.onHideNotice.bind(null, item.id)}
                  onTouchEnd={this.onHideNotice.bind(null, item.id)}
                />
              </div>
            </div>
          );
        })}
      </div>
    );
  }
}
