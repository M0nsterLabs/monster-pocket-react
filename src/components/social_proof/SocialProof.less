.notice-wrapper {
  position: absolute;
  right: 20px;
  * {
    font-family: 'PT Sans', sans-serif;
  }
  .notice {
    margin-top: 20px;
    box-sizing: border-box;
    padding-top: 20px;
    padding-bottom: 20px;
    padding-left: 50px;
    padding-right: 50px;
    border-radius: 3px;
    position: relative;
    transition: margin .5s cubic-bezier(0.68, -0.55, 0.265, 1.55);
    -webkit-transition: margin .5s cubic-bezier(0.68, -0.55, 0.265, 1.55);
    -moz-transition: margin .5s cubic-bezier(0.68, -0.55, 0.265, 1.55);
    .notice-text{
      &.t3{
        color: #ffffff;
      }
    }
    &_show {
      .animation-notice(show_notice);
    }
    &_hide {
      .animation-notice(hide_notice);
    }
    &_type {
      &-default {
        background-color: rgba(84, 110, 121, 0.9);
        .notice-shadow(rgba(35, 50, 56, 0.2))
      }
      &-success {
        background-color: rgba(25, 183, 67, 0.9);
        .notice-shadow(rgba(0, 96, 35, 0.2));
      }
    }
    &_has-icon {
      padding-left: 60px;
    }
    &_icon {
      &-flag {

      }
    }

    &__closeBlock {
      background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAiIGhlaWdodD0iMTAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CiA8Zz4KICA8dGl0bGU+YmFja2dyb3VuZDwvdGl0bGU+CiAgPHJlY3QgeD0iLTEiIHk9Ii0xIiB3aWR0aD0iMTIiIGhlaWdodD0iMTIiIGlkPSJjYW52YXNfYmFja2dyb3VuZCIgZmlsbD0ibm9uZSIvPgogPC9nPgoKIDxnPgogIDx0aXRsZT5MYXllciAxPC90aXRsZT4KICA8cGF0aCBzdHJva2U9Im51bGwiIGQ9Im05LjcwMDIzLDEuMzQ1NjdjMC4xNDM5LDAuMTQzNCAwLjE0MzksMC4zNzg0NCAwLDAuNTIxODNsLTIuODg1NCwyLjg3MjYyYy0wLjE0MzksMC4xNDI5NSAtMC4xNDM5LDAuMzc4NDQgMCwwLjUyMjNsMi44ODQwMiwyLjg3MzU1YzAuMTQzOSwwLjE0MzQgMC4xNDM5LDAuMzc4NDQgMCwwLjUyMjNsLTEuMDUwNTMsMS4wNDMyMmMtMC4xNDQzNiwwLjE0Mjk1IC0wLjM4MDY4LDAuMTQyOTUgLTAuNTI1MDQsMGwtMi44ODMxLC0yLjg3MzA5Yy0wLjE0MzksLTAuMTQzNCAtMC4zODA2OCwtMC4xNDM0IC0wLjUyNDU4LDBsLTIuODg1NCwyLjg3MTI1Yy0wLjE0MzksMC4xNDM0IC0wLjM4MDY4LDAuMTQzNCAtMC41MjQ1OCwwbC0xLjA0OTE1LC0xLjA0NDU5Yy0wLjE0MzksLTAuMTQzNCAtMC4xNDM5LC0wLjM3ODQ0IDAsLTAuNTIyM2wyLjg4NjMyLC0yLjg3MjE3YzAuMTQzOSwtMC4xNDI5NSAwLjE0MzksLTAuMzc4NDQgMCwtMC41MjIzbC0yLjg4NDAyLC0yLjg3NGMtMC4xNDQzNiwtMC4xNDM0IC0wLjE0NDM2LC0wLjM3ODkgMCwtMC41MjIzbDEuMDUwMDcsLTEuMDQzNjhjMC4xNDQ4MiwtMC4xNDI5NSAwLjM4MDY4LC0wLjE0Mjk1IDAuNTI1MDQsMC4wMDA0NmwyLjg4MjY1LDIuODcyNjJjMC4xNDQzNiwwLjE0Mzg2IDAuMzgwNjgsMC4xNDM4NiAwLjUyNDU4LDAuMDAwNDZsMi44ODU0LC0yLjg3MTI1YzAuMTQzNDQsLTAuMTQyOTUgMC4zNzk3NiwtMC4xNDI5NSAwLjUyNDU4LDBsMS4wNDkxNSwxLjA0NTA2eiIgaWQ9InN2Z18xIiBmaWxsPSIjZmZmZmZmIi8+CiA8L2c+Cjwvc3ZnPgo=);
      cursor: pointer;
      pointer-events: auto;
      margin-left: 20px;
      align-self: flex-start;
      position: absolute;
      right: 20px;
      top: 20px;
      width: 10px;
      height: 10px;
      &__closeArea {
        cursor: pointer;
        position: absolute;
        width: 40px;
        height: 40px;
        left: -15px;
        top: -15px;
      }
    }
  }
}

.notice-shadow(@color) {
  -webkit-box-shadow: 0px 5px 10px 0px @color;
  -moz-box-shadow: 0px 5px 10px 0px @color;
  box-shadow: 0px 5px 10px 0px @color;
}

.animation-notice(@name) {
  @time: .5s;
  @timingFunction: cubic-bezier(0.68, -0.55, 0.265, 1.55);

  animation-duration: @time;
  -webkit-animation-duration: @time;
  -moz-animation-duration: @time;

  animation-fill-mode: forwards;
  -webkit-animation-fill-mode: forwards;
  -moz-animation-fill-mode: forwards;

  animation-iteration-count: 1;
  -webkit-animation-iteration-count: 1;
  -moz-animation-iteration-count: 1;

  animation-name: @name;
  -webkit-animation-name: @name;
  -moz-animation-name: @name;

  -webkit-animation-timing-function: @timingFunction;
  animation-timing-function: @timingFunction;
  -moz-animation-timing-function: @timingFunction;

  //@top_start: -100px;
  @top_finish: 20px;

  @-webkit-keyframes show_notice {
    0% {
      //margin-top: @top_start;
      opacity: 0;
    }
    100% {
      margin-top: @top_finish;
      opacity: 1;
    }
  }
  @-ms-keyframes show_notice {
    0% {
      //margin-top: @top_start;
      opacity: 0;
    }
    100% {
      margin-top: @top_finish;
      opacity: 1;
    }
  }
  @-moz-keyframes show_notice {
    0% {
      //margin-top: @top_start;
      opacity: 0;
    }
    100% {
      margin-top: @top_finish;
      opacity: 1;
    }
  }
  @keyframes show_notice {
    0% {
      //margin-top: @top_start;
      opacity: 0;
    }
    100% {
      margin-top: @top_finish;
      opacity: 1;
    }
  }

  @-webkit-keyframes hide_notice {
    0% {
      transform: scale(1);
      opacity: 1;
    }
    100% {
      transform: scale(0.8);
      opacity: 0;
    }
  }
  @-ms-keyframes hide_notice {
    0% {
      transform: scale(1);
      opacity: 1;
    }
    100% {
      transform: scale(0.8);
      opacity: 0;
    }
  }
  @-moz-keyframes hide_notice {
    0% {
      transform: scale(1);
      opacity: 1;
    }
    100% {
      transform: scale(0.8);
      opacity: 0;
    }
  }
  @keyframes hide_notice {
    0% {
      transform: scale(1);
      opacity: 1;
    }
    100% {
      transform: scale(0.8);
      opacity: 0;
    }
  }
}