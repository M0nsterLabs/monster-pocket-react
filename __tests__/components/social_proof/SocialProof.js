/* global describe: false, it: false, expect: false, toMatchSnapshot: false, beforeEach: false, afterEach: false, jest: false */
import React from 'react';
import ReactDOM from 'react-dom';
import {
  mount,
  shallow,
  render
} from 'enzyme';
import SocialProof from '../../../src/components/social_proof/SocialProof.jsx'


let container;
let animationend;
beforeEach(() => {
  jest.useFakeTimers();
  jest.mock('shortid');
  container = document.createElement('div');
  animationend = new Event('animationend');
});

describe('SocialProofs:', () => {
  test('Only text params', () => {
    let component = mount(<SocialProof notice={[{text : 'Only text params'}]}/>);
    expect(component.html()).toMatchSnapshot();
  });

  test('Notice with icon & text params', () => {
    let component = mount(<SocialProof notice={[{
      text : 'Notice with icon & text params',
      icon : 'head'
    }]}/>);
    expect(component.html()).toMatchSnapshot();
  });

  test('Only text params & className', () => {
    let component = mount(<SocialProof notice={[{
      text      : 'Only text params & className',
      className : 'test'
    }]}/>);
    expect(component.html()).toMatchSnapshot();
  });

  test('Only text params & re-render component with object', () => {
    let component = mount(<SocialProof notice={[{
      text      : 'Only text params',
      className : 'test'
    }]}/>);
    component.setProps({
      notice : {
        text      : 're-render component with object',
        className : 'test'
      }
    });
    expect(component.html()).toMatchSnapshot();
  });

  test('Only text params & re-render component with array', () => {
    let component = mount(<SocialProof notice={[{
      text      : 'Only text params',
      className : 'test'
    }]}/>);
    component.setProps({
      notice : [{
        text      : 're-render component with array',
        className : 'test'
      }]
    });
    expect(component.html()).toMatchSnapshot();
  });

  test('Open notice & click the close', () => {
    let component = mount(<SocialProof notice={
      {
        text      : 'Only text params',
        className : 'test'
      }
    }/>);
    component.find('.notice__closeBlock__closeArea').simulate('click');
    jest.runAllTimers();
    component.find('.notice_hide').node.dispatchEvent(animationend);
    expect(component.find('.notice').exists()).toBe(false);
  });

  test('Open notices & click the close of one', () => {
    let component = mount(<SocialProof notice={[
      {
        text      : 'Only text params',
        className : 'test'
      },
      {
        text      : 'Only text params copy',
        className : 'test'
      }
    ]}/>);
    component.find('.notice__closeBlock__closeArea').last().simulate('click');
    jest.runAllTimers();
    component.find('.notice_hide').node.dispatchEvent(animationend);
    expect(component.html()).toMatchSnapshot();
  });

});