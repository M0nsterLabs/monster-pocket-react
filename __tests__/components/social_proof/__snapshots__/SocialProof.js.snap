exports[`SocialProofs: Notice with icon & text params 1`] = `"<div class=\"notice-wrapper \" style=\"width: 340px;\"><div class=\"notice notice_type-default notice_has-icon notice_icon notice_icon_head   notice_show \" data-id=\"qwerty1\" style=\"margin-top: 0px;\"><span class=\"notice-text t3\">Notice with icon &amp; text params</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div></div>"`;

exports[`SocialProofs: Only text params & className 1`] = `"<div class=\"notice-wrapper \" style=\"width: 340px;\"><div class=\"notice notice_type-default   test notice_show \" data-id=\"qwerty2\" style=\"margin-top: 0px;\"><span class=\"notice-text t3\">Only text params &amp; className</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div></div>"`;

exports[`SocialProofs: Only text params & re-render component with array 1`] = `"<div class=\"notice-wrapper \" style=\"width: 340px;\"><div class=\"notice notice_type-default notice_show\" data-id=\"qwerty6\" style=\"margin-top: -20px;\"><span class=\"notice-text t3\">re-render component with array</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div><div class=\"notice notice_type-default    \" data-id=\"qwerty5\" style=\"margin-top: 20px;\"><span class=\"notice-text t3\">Only text params</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div></div>"`;

exports[`SocialProofs: Only text params & re-render component with object 1`] = `"<div class=\"notice-wrapper \" style=\"width: 340px;\"><div class=\"notice notice_type-default notice_show\" data-id=\"qwerty4\" style=\"margin-top: -20px;\"><span class=\"notice-text t3\">re-render component with object</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div><div class=\"notice notice_type-default    \" data-id=\"qwerty3\" style=\"margin-top: 20px;\"><span class=\"notice-text t3\">Only text params</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div></div>"`;

exports[`SocialProofs: Only text params 1`] = `"<div class=\"notice-wrapper \" style=\"width: 340px;\"><div class=\"notice notice_type-default    notice_show \" data-id=\"qwerty0\" style=\"margin-top: 0px;\"><span class=\"notice-text t3\">Only text params</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div></div>"`;

exports[`SocialProofs: Open notices & click the close of one 1`] = `"<div class=\"notice-wrapper \" style=\"width: 340px;\"><div class=\"notice notice_type-default    notice_transition\" data-id=\"qwerty8\" style=\"margin-top: 0px;\"><span class=\"notice-text t3\">Only text params</span><div class=\"notice__closeBlock\"><div class=\"notice__closeBlock__closeArea\"></div></div></div></div>"`;
